tests/integration.rs

```diff
@@ -0,0 +1,433 @@
+    // respectively. If unset "./pebble" and "./pebble-challtestsrv" are used.
+    //
+    // Returns only once the Pebble CA server interface is responding.
+    fn new(config: Config) -> io::Result<Self> {
+        let config_file = NamedTempFile::new()?;
+        fs::write(&config_file, serde_json::to_string_pretty(&config)?)?;
+
+        let pebble_path = env::var("PEBBLE").ok().unwrap_or("./pebble".to_string());
+        let challtestsrv_path = env::var("CHALLTESTSRV")
+            .ok()
+            .unwrap_or("./pebble-challtestsrv".to_string());
+
+        let pebble = Subprocess::new(
+            Command::new(&pebble_path)
+                .arg("-config")
+                .arg(config_file.path())
+                .arg("-dnsserver")
+                .arg("127.0.0.1:8053") // Matched to default -dns01 addr for pebble-challtestsrv.
+                .arg("-strict"),
+        )?;
+
+        let challtestsrv = Subprocess::new(
+            Command::new(&challtestsrv_path)
+                .arg("-doh-cert")
+                .arg("tests/testdata/server.pem")
+                .arg("-doh-cert-key")
+                .arg("tests/testdata/server.key"),
+        )?;
+
+        wait_for_server(&config.pebble.listen_address);
+
+        Ok(Self {
+            config,
+            config_file,
+            pebble,
+            challtestsrv,
+        })
+    }
+
+    // Return an HTTP client configured to trust the Pebble management interface root CA.
+    //
+    // Note: this is distinct from the CA issuer.
+    fn http_client(&self) -> HyperClient<hyper_rustls::HttpsConnector<HttpConnector>, Full<Bytes>> {
+        let mut roots = RootCertStore::empty();
+        roots.add_parsable_certificates(
+            CertificateDer::pem_file_iter("tests/testdata/ca.pem")
+                .unwrap()
+                .map(|result| result.unwrap()),
+        );
+
+        HyperClient::builder(TokioExecutor::new()).build(
+            hyper_rustls::HttpsConnectorBuilder::new()
+                .with_tls_config(
+                    rustls::ClientConfig::builder()
+                        .with_root_certificates(roots)
+                        .with_no_client_auth(),
+                )
+                .https_or_http()
+                .enable_http1()
+                .enable_http2()
+                .build(),
+        )
+    }
+
+    // Provision an HTTP-01 challenge response for the given token and key authorization.
+    //
+    // The Pebble challenge test server will be configured to respond to HTTP-01 challenge
+    // requests for the provided token by returning the expected key auth.
+    async fn add_http01_response(
+        &self,
+        token: &str,
+        key_auth: &str,
+    ) -> Result<(), Box<dyn StdError>> {
+        let client = self.http_client();
+
+        let body = json!({
+            "token": token,
+            "content": key_auth,
+        })
+        .to_string();
```

Comment:
How do you feel about defining an actual `Serialize`-implementing type for this instead of using `json!()`?

Maybe serialize directly into a `Bytes` instead of converting via `String`?

Reply:
Done x2.